### PR TITLE
feat: add simple pdf extraction service

### PR DESCRIPTION
**Pull request by tangopapatime**
# Description

# Description

Added a local pdf extraction service in Python/Flask
Added tera-drag-n-drop component to the project - v0 of setting up for user uploading.

Resolves #(issue)
Issue 1020

[import_resource.webm](https://user-images.githubusercontent.com/98172806/236302229-168baec6-2614-44a9-970e-d7cbf23c1019.webm)

### CODE REVIEW

**Comment by tangopapatime** [comment]
I haven't yet resolved the multipart pass through issue but have found a workaround for now.  Need to add additional documentation as well.

File: packages/services/hmi-server/src/main/java/software/uncharted/terarium/hmiserver/resources/pdfextractionservice/PDFExtractionResource.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+		return extractionProxy.convertPDFTask(body);
+
+	}
+
+	// checks a taskId status
+	@GET
+	@Path("/task_result/{taskId}")
+	public Response getTaskStatus(
+			@PathParam("taskId") final String taskId) {
+		return extractionProxy.getTaskStatus(taskId);
```
Comment (dvince2) [inline packages/services/hmi-server/src/main/java/software/uncharted/terarium/hmiserver/resources/pdfextractionservice/PDFExtractionResource.java:63]: Re: your recent change to take `PDFExtractionResponseType` out of `Types.ts`...

If you create a java class for `PDFExtractionResponseType` and then annotate it with `@TSModel` you will get a POJO created for the back end to use and then the ts implementation would come for free :)

File: packages/client/hmi-client/src/types/common.ts [hunk 2]
```diff
@@ -87,3 +87,15 @@ export type Tab = {
...
 };
+
+export enum AcceptedTypes {
+	PDF = 'application/pdf',
+	JPG = 'image/jpg',
+	JPEG = 'image/jpeg',
+	PNG = 'image/png'
+}
+
+export interface PDFExtractionResponseType {
```
Comment (dvince2) [inline packages/client/hmi-client/src/types/common.ts:98]: Re: your recent change to take PDFExtractionResponseType out of Types.ts...

If you create a java class for PDFExtractionResponseType and then annotate it with @TSModel you will get a POJO created for the back end to use and then the ts implementation would come for free :)

File: packages/services/hmi-server/src/main/java/software/uncharted/terarium/hmiserver/proxies/pdfextractionservice/PDFExtractionServiceProxy.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+			@DefaultValue("pypdf2") @QueryParam("extraction_method") String extractionMethod,
+			@DefaultValue("true") @QueryParam("extract_images") String extractImages);
+
+	@POST
+	@Path("/convertpdftask")
+	@Consumes(MediaType.MULTIPART_FORM_DATA)
+	Response convertPDFTask(@MultipartForm PDFExtractionMultipartBody formData);
+
+	@GET
+	@Path("/task_result/{taskId}")
```
Comment (dvince2) [inline packages/services/hmi-server/src/main/java/software/uncharted/terarium/hmiserver/proxies/pdfextractionservice/PDFExtractionServiceProxy.java:30]: endpoint should use hyphen, not underscore https://github.com/DARPA-ASKEM/Terarium/issues/862

File: packages/services/hmi-server/src/main/java/software/uncharted/terarium/hmiserver/resources/DownloadResource.java [hunk 4]
```diff
@@ -94,15 +96,82 @@ private byte[] getPDF(final String url) throws IOException, URISyntaxException {
...
 
+	/**
+	 * Gets a PDF file from a given url
+	 *
+	 * @param url the url location (that may contain redirects)
+	 * @return the pdf file
+	 * @throws IOException
+	 * @throws URISyntaxException
+	 */
+	private String getPDFURl(final String url) throws IOException, URISyntaxException {
```
Comment (dvince2) [inline packages/services/hmi-server/src/main/java/software/uncharted/terarium/hmiserver/resources/DownloadResource.java:107]: `L`

File: packages/client/hmi-client/src/types/common.ts (hunk 2 above)
Comment (tangopapatime) [reply packages/client/hmi-client/src/types/common.ts:98]: Thanks @dvince2 .  Yeah that was my original thought but I'm not quite sure this actually belongs so I just kept it simple!

File: packages/services/hmi-server/src/main/java/software/uncharted/terarium/hmiserver/resources/DownloadResource.java [hunk 5]
```diff
@@ -94,15 +96,82 @@ private byte[] getPDF(final String url) throws IOException, URISyntaxException {
...
+			}
+		}
+		return url;
+	}
+
 	@GET
 	@Produces(MediaType.APPLICATION_OCTET_STREAM)
 	public Response get(@QueryParam("doi") final String doi) throws IOException, URISyntaxException {
-
 		final byte[] pdfBytes = getPDF("https://unpaywall.org/" + doi);
```
Comment (dvince2) [inline packages/services/hmi-server/src/main/java/software/uncharted/terarium/hmiserver/resources/DownloadResource.java:156]: not required at all, but, if you ever want to do any sort of validation on a doi there is a regex expression used in ExtractionResource called `DOI_VALIDATION_PATTERN`

**Comment by tangopapatime** [comment]
The major outstanding follow up task is to integrate this service as part of the orchestration.  Especially if actually want this to work on `staging`.  Definitely not my forte but can give it a shot if someone can outline the major steps so that I'm not forgetting anything.

**Review by dvince2 — APPROVED** [review]
Java side looks good - just a few small comments on it to do before this is finalized.  I took a quick look at the front end stuff but someone else may want to comment there, and I'm not sure about python 😂

File: packages/services/hmi-server/src/main/java/software/uncharted/terarium/hmiserver/resources/DownloadResource.java (hunk 4 above)
Comment (tangopapatime) [reply packages/services/hmi-server/src/main/java/software/uncharted/terarium/hmiserver/resources/DownloadResource.java:107]: or R? :P

Comment (dvince2) [reply packages/services/hmi-server/src/main/java/software/uncharted/terarium/hmiserver/resources/DownloadResource.java:107]: `private String gEtPdFuRl(final String UrL)`

**Comment by tangopapatime** [comment]
I still need an check @YohannParis . Take a peak at this and see if it's a-ok.

**Review by YohannParis — APPROVED** [review]

**Comment by YohannParis** [comment]
> I still need an check @YohannParis . Take a peak at this and see if it's a-ok.

After a _careful_ review of those 1000+ lines of code, I approve.

**Pull request merged**